File: src/lib/worker/summarize-conversation.ts
/**
 * Short-term memory: condense old turns of a conversation into a one-paragraph
 * summary on `conversations.summary_md`, so long chats survive token compaction
 * without losing context.
 *
 * Triggered after each agent turn (fire-and-forget). Idempotent and
 * rate-limited: only re-summarizes if message_count grew by >= TRIGGER_EVERY
 * since the last summary.
 *
 * Cost note: one short LLM call per summary, gated by the trigger threshold.
 */
import { generateText } from "ai";
import { catalogPool } from "../catalog/db";
import { setConversationSummary } from "../catalog/memories";
import { docWriterModel } from "./openrouter";
import { log } from "../shared/log";

const TRIGGER_EVERY = 4;
const SYSTEM = `You are summarizing the older portion of a chat between a
data analyst (the user) and an AI agent (Loom) so the agent can carry
forward context without re-reading every old message.

Write ONE compact Markdown paragraph (max ~120 words) covering:
- What the user is investigating (their goal in this chat)
- Key decisions / constraints they've established
- Tables, views, skills, or domains the agent has used so far
- Open threads that may need follow-up

Be terse. Use specifics from the messages, not generic phrases. Skip
greetings, acknowledgements, and tool-call mechanics. Don't repeat
information already in the system prompt.`;

export async function maybeSummarizeConversation(slug: string): Promise<void> {
  const r = await catalogPool.query<{
    total: number;
    last_summarized: number;
    last_user_text: string | null;
  }>(
    `SELECT
       (SELECT count(*)::int FROM messages WHERE conversation_id = $1) AS total,
       COALESCE(c.summarized_turn_count, 0) AS last_summarized,
       (SELECT m.parts::text FROM messages m
         WHERE m.conversation_id = $1 AND m.role = 'user'
         ORDER BY m.ord DESC LIMIT 1) AS last_user_text
       FROM conversations c WHERE c.slug = $1`,
    [slug],
  );
  const row = r.rows[0];
  if (!row) return;
  if (row.total - row.last_summarized < TRIGGER_EVERY) return;

  // Pull every message except the very latest pair (keep them verbatim so
  // they read naturally in the new turn).
  const messagesRes = await catalogPool.query<{
    role: string;
    parts: unknown;
    ord: number;
  }>(
    `SELECT role, parts, ord
       FROM messages
      WHERE conversation_id = $1
      ORDER BY ord ASC`,
    [slug],
  );
  if (messagesRes.rows.length === 0) return;

  // Flatten parts into plain text for the summarizer prompt.
  const transcript = messagesRes.rows
    .map((m) => `### ${m.role.toUpperCase()}\n${extractText(m.parts)}`)
    .join("\n\n");
  if (transcript.length < 200) return; // not enough material

  try {
    const out = await generateText({
      model: docWriterModel(),
      system: SYSTEM,
      prompt: transcript.slice(0, 18_000), // cap input to ~4k tokens
      maxRetries: 1,
    });
    const summary = out.text.trim();
    if (summary.length === 0) return;
    await setConversationSummary(slug, summary, row.total);
    log.info("conv_summary.updated", { slug, turn_count: row.total, bytes: summary.length });
  } catch (e) {
    log.warn("conv_summary.failed", { slug, err: String(e) });
  }
}

function extractText(parts: unknown): string {
  if (!Array.isArray(parts)) return "";
  return parts
    .map((p) => {
      const obj = p as { type?: string; text?: string; toolName?: string; output?: unknown };
      if (obj.type === "text" && typeof obj.text === "string") return obj.text;
      if (typeof obj.type === "string" && obj.type.startsWith("tool-")) {
        const name = obj.type.replace(/^tool-/, "");
        const out = obj.output ? JSON.stringify(obj.output).slice(0, 200) : "";
        return `[tool ${name}${out ? `: ${out}` : ""}]`;
      }
      return "";
    })
    .filter(Boolean)
    .join("\n");
}

File: backend/app/workers/summarize_conversation.py
"""Per-conversation short-term memory.

Port of src/lib/worker/summarize-conversation.ts. Condenses older turns into
a one-paragraph summary on conversations.summary_md so long chats survive
token compaction. Triggered fire-and-forget after each agent turn; only
re-runs when total messages grew by TRIGGER_EVERY since the last summary.
"""
from __future__ import annotations

import json
import logging
from typing import Any

from psycopg.rows import dict_row

from ..catalog.memories import set_conversation_summary
from ..db import get_conn
from ..llm import async_client, chat_model_id

log = logging.getLogger(__name__)

_TRIGGER_EVERY = 4

SYSTEM = """You are summarizing the older portion of a chat between a
data analyst (the user) and an AI agent (Loom) so the agent can carry
forward context without re-reading every old message.

Write ONE compact Markdown paragraph (max ~120 words) covering:
- What the user is investigating (their goal in this chat)
- Key decisions / constraints they've established
- Tables, views, skills, or domains the agent has used so far
- Open threads that may need follow-up

Be terse. Use specifics from the messages, not generic phrases. Skip
greetings, acknowledgements, and tool-call mechanics."""


async def maybe_summarize_conversation(slug: str) -> None:
    async with get_conn() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(
                """
                SELECT
                  (SELECT count(*)::int FROM messages WHERE conversation_id = %s) AS total,
                  COALESCE(c.summarized_turn_count, 0) AS last_summarized
                  FROM conversations c WHERE c.slug = %s
                """,
                (slug, slug),
            )
            row = await cur.fetchone()
    if not row:
        return
    total = int(row["total"] or 0)
    last = int(row["last_summarized"] or 0)
    if total - last < _TRIGGER_EVERY:
        return

    async with get_conn() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(
                """
                SELECT role, parts, ord
                  FROM messages
                 WHERE conversation_id = %s
                 ORDER BY ord ASC
                """,
                (slug,),
            )
            messages = await cur.fetchall()
    if not messages:
        return

    transcript = "\n\n".join(
        f"### {m['role'].upper()}\n{_extract_text(m['parts'])}" for m in messages
    )
    if len(transcript) < 200:
        return

    try:
        resp = await async_client().chat.completions.create(
            model=chat_model_id(),
            messages=[
                {"role": "system", "content": SYSTEM},
                {"role": "user", "content": transcript[:18_000]},
            ],
            temperature=0.2,
        )
        text = (resp.choices[0].message.content or "").strip() if resp.choices else ""
    except Exception as e:  # noqa: BLE001
        log.warning("conv_summary.failed slug=%s err=%s", slug, e)
        return
    if not text:
        return
    await set_conversation_summary(slug, text, total)
    log.info("conv_summary.updated slug=%s turn_count=%s bytes=%s", slug, total, len(text))


def _extract_text(parts: Any) -> str:
    if not isinstance(parts, list):
        return ""
    out: list[str] = []
    for p in parts:
        if not isinstance(p, dict):
            continue
        t = p.get("type")
        if t == "text" and isinstance(p.get("text"), str):
            out.append(p["text"])
        elif isinstance(t, str) and t.startswith("tool-"):
            name = t.removeprefix("tool-")
            output = p.get("output")
            tail = f": {json.dumps(output)[:200]}" if output is not None else ""
            out.append(f"[tool {name}{tail}]")
    return "\n".join(out)
